### PR TITLE
[Perf] Add healthy-path regression benchmarks

### LINKED ISSUE #23: [Feature] Add reproducible performance benchmarks and regression guards for healthy-path overhead
## Problem

Performance is part of the project's correctness envelope, and the README makes concrete healthy-path overhead claims. `docs/gemini.md` contains useful point-in-time GEMINI measurements and `docs/validation.md` includes overhead smoke evidence, but the repository has no checked-in benchmark harness or automated regression threshold that can reproduce those measurements systematically and detect regressions over time. SCOUT's README-level overhead claim also needs clearly traceable current-package evidence.

## Proposed solution

Add a reproducible benchmark suite covering:

- baseline vs GEMINI-only vs SCOUT-only vs combined;
- representative checkpoint cadence and sizes;
- pinned-memory and replication overhead;
- replay sampling cadence;
- single-host and, when available, two-host communication interference;
- throughput, p50/p95 step latency, and peak host/GPU memory.

Run a small stable benchmark tier on scheduled hardware, store machine-readable results, and compare against an explicit baseline/tolerance.

## Alternatives considered

- Keep performance validation entirely manual: flexible, but regressions are discovered late and historical claims are difficult to reproduce.
- Run noisy large-scale benchmarks on every PR: too expensive and too variable; a scheduled stable tier is preferable.

## Framework and topology scope

Start with native PyTorch and representative single-host GPU topology; expand to framework-specific and two-host cases where behavior differs materially. Benchmark methodology should record DP/TP/PP/CP/SP/EP dimensions when applicable.

## Suggested validation

- Reproduce the existing GEMINI point-in-time measurements within expected variance on a documented environment.
- Deliberately add a synchronization or blocking operation and verify the regression guard detects it.
- Confirm result records contain commit SHA, package/framework versions, hardware, topology, workload, seed, and metrics.

## Additional context

Existing measurements should be preserved and clearly labeled as historical/current-package evidence rather than removed.

## Before submitting

- [x] I searched open and closed issues for related proposals.
- [x] I described the problem independently of the proposed implementation.

### PR DESCRIPTION
**Pull request by zhuangwang93**
Closes #23.

## What changed
- adds isolated native-PyTorch baseline, GEMINI-only, SCOUT-only, and combined benchmark modes;
- records revision, environment, topology, workload, protection settings, throughput, p50/p95 latency, peak host/GPU memory, and per-rank data;
- adds explicit throughput/latency regression thresholds, machine-readable summaries, commands, and SHA-256 checksums;
- runs the enforced two-GPU tier in the scheduled/manual GPU qualification workflow;
- labels existing README measurements as historical workload-specific evidence and documents qualification boundaries.

## Validation
- `python -m pytest tests/unit/test_healthy_path_benchmark.py -q` — 4 passed
- sandbox-compatible CPU suite — 602 passed
- Ruff and pre-commit — passed
- CPU baseline/GEMINI harness smoke — passed

The SCOUT/combined path requires CUDA and remains to be exercised by the trusted two-GPU qualification runner. That workflow is scheduled/manual rather than pull-request-triggered so untrusted code does not execute on the self-hosted runner. Existing actionlint warnings for the repository's custom `gpu` and `lm-resiliency` labels are unchanged.

### CODE REVIEW

File: benchmarks/healthy_path.py [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+        handle = enable_resiliency(
+            model,
+            optimizer,
+            interval=args.interval,
+            enable_checkpoint=enable_checkpoint,
+            enable_detection=enable_detection,
+            checkpoint=InMemoryCkptConfig(
+                enable=enable_checkpoint,
+                interval=args.interval,
+                replication_jump=max(1, world_size // 2),
```
[the remark this answers is not in the document]
Comment (lmresiliencydev) [reply benchmarks/healthy_path.py]: Fixed in 90fc28b. The controller rejects odd multi-rank world sizes whenever GEMINI or combined mode is requested, and the worker uses the same validated replication helper defensively. Baseline/SCOUT-only odd topologies remain supported.

File: benchmarks/healthy_path.py [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+        return sum(
+            _tree_tensor_bytes(key) + _tree_tensor_bytes(item) for key, item in value.items()
+        )
+    if isinstance(value, (list, tuple)):
+        return sum(_tree_tensor_bytes(item) for item in value)
+    return 0
+
+
+def _max_rss_bytes() -> int:
+    value = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
```
[the remark this answers is not in the document]
Comment (lmresiliencydev) [reply benchmarks/healthy_path.py:174]: Fixed in 90fc28b. Each SCOUT rank now reads the live OOB daemon VmHWM from /proc and reports parent peak RSS, daemon peak RSS, and their conservative combined total separately. The methodology documents the component accounting.

File: benchmarks/healthy_path.py [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+                torch.cuda.max_memory_allocated(device) if device.type == "cuda" else None
+            ),
+        }
+        rank_results: list[dict[str, Any] | None] = [None] * world_size
+        dist.all_gather_object(rank_results, rank_result)
+
+        if rank != 0:
+            return {}
+        complete_results = [result for result in rank_results if result is not None]
+        all_step_times = [value for result in complete_results for value in result["step_times_ms"]]
```
[the remark this answers is not in the document]
Comment (lmresiliencydev) [reply benchmarks/healthy_path.py]: Fixed in 90fc28b. Rank timings are now collapsed by step offset using the maximum rank latency before p50/p95 calculation and threshold enforcement. The job-level samples are also retained in the result, with a regression test where rank-local outliers would be diluted by flattening.

File: benchmarks/healthy_path.py [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+            device=device,
+        )
+
+    try:
+        for step in range(args.warmup_steps):
+            _train_step(model, optimizer, rank, step, args, device)
+
+        if handle is not None and handle.ckpt_manager is not None:
+            handle.ckpt_manager.maybe_wait()
+        dist.barrier()
```
[the remark this answers is not in the document]
Comment (lmresiliencydev) [reply benchmarks/healthy_path.py:309]: Fixed in 9789ae7. OOBHangService now exposes a multiprocessing readiness event set only after the child joins its independent Gloo group and constructs the detector. SCOUT/combined benchmarks wait for that signal before the rank barrier and memory/timing reset. Unit tests cover ready, timeout, and early-exit paths.

File: benchmarks/healthy_path.py [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+                "hidden_size": args.hidden_size,
+                "layers": args.layers,
+                "heads": args.heads,
+                "checkpoint_bytes_per_rank": _tree_tensor_bytes(model.state_dict())
+                + _tree_tensor_bytes(optimizer.state_dict()),
+            },
+            "protection": {
+                "checkpoint_interval": args.interval if enable_checkpoint else None,
+                "replay_interval": args.interval if enable_detection else None,
+                "pin_memory": args.pin_memory if enable_checkpoint else None,
```
[the remark this answers is not in the document]
Comment (lmresiliencydev) [reply benchmarks/healthy_path.py]: Fixed in 9789ae7. The effective device-gated pin-memory value is computed once, passed to InMemoryCkptConfig, and recorded in the evidence. The CPU GEMINI smoke confirms the artifact now reports false.

File: benchmarks/README.md [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+For a fast CPU harness smoke without a performance claim:
+
+```bash
+python benchmarks/run_healthy_path.py \
+  --device cpu \
+  --world-size 1 \
+  --modes baseline gemini \
+  --steps 3 \
+  --warmup-steps 1 \
```
[the remark this answers is not in the document]
Comment (lmresiliencydev) [reply benchmarks/README.md:27]: Fixed in 9789ae7. The documented four-step CPU smoke now sets --interval 1, so GEMINI exercises checkpoint capture/copy/manager paths. I reran that exact smoke locally and confirmed the interval and effective pinning fields.

File: benchmarks/healthy_path.py [hunk 7]
```diff
@@ -0,0 +1,456 @@
...
+            "commit_sha": os.environ.get("GITHUB_SHA") or _git_revision(),
+            "mode": args.mode,
+            "environment": {
+                "host": platform.node(),
+                "platform": platform.platform(),
+                "python": platform.python_version(),
+                "torch": torch.__version__,
+                "cuda_runtime": torch.version.cuda,
+                "device": args.device,
+                "device_name": torch.cuda.get_device_name(0) if device.type == "cuda" else None,
```
[the remark this answers is not in the document]
Comment (lmresiliencydev) [reply benchmarks/healthy_path.py]: Fixed in 9789ae7. Each rank now records the model for its actual local CUDA device. Rank 0 gathers the ordered device list and rejects qualification when the models differ; both the per-rank records and environment topology retain the device identities.

File: benchmarks/healthy_path.py [hunk 8]
```diff
@@ -0,0 +1,456 @@
...
+
+def _oob_daemon_peak_rss_bytes(handle: Any) -> int | None:
+    replay_harness = getattr(handle, "replay_harness", None)
+    service = getattr(replay_harness, "_oob_service", None)
+    return _process_peak_rss_bytes(getattr(service, "pid", None))
+
+
+def _git_revision() -> str | None:
+    completed = subprocess.run(
+        ["git", "rev-parse", "HEAD"],
```
[the remark this answers is not in the document]
Comment (lmresiliencydev) [reply benchmarks/healthy_path.py:211]: Fixed in 9789ae7. The controller captures git porcelain state before producing any outputs, passes that immutable dirty flag to every isolated mode, and records worktree_dirty beside commit_sha. Direct per-mode runs also inspect their worktree.

File: benchmarks/run_healthy_path.py [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+def summarize_results(
+    runs: dict[str, dict[str, Any]], thresholds: dict[str, Any]
+) -> dict[str, Any]:
+    baseline = runs["baseline"]["metrics"]
+    comparisons = []
+    violations = []
+    for metric, policy in thresholds["metrics"].items():
+        for mode, limit in policy["maximum_regression_percent"].items():
+            if mode not in runs:
+                continue
```
[the remark this answers is not in the document]
Comment (lmresiliencydev) [reply benchmarks/run_healthy_path.py:58]: Fixed in 9789ae7. --enforce now rejects a mode set with no threshold-protected mode before launching torchrun, and summaries with no comparisons are explicitly not_qualified rather than passed. The baseline-only enforced command exits with argparse status 2 in local validation.

File: benchmarks/run_healthy_path.py [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+            f"{comparison['regression_percent']:.2f}% | "
+            f"{comparison['maximum_regression_percent']:.2f}% | "
+            f"{'passed' if comparison['passed'] else 'failed'} |"
+        )
+    path.write_text("\n".join(lines) + "\n", encoding="utf-8")
+
+
+def _write_checksums(output_dir: Path) -> None:
+    lines = []
+    for path in sorted(item for item in output_dir.iterdir() if item.is_file()):
```
[the remark this answers is not in the document]
Comment (lmresiliencydev) [reply benchmarks/run_healthy_path.py]: Fixed in 9789ae7. The controller removes only known benchmark outputs at invocation start and checksums an explicit set of files generated by the current mode set. Tests verify stale files are excluded and unrelated user files are preserved.

**Review by lmresiliencydev — APPROVED** [review]
